### PR TITLE
feat(FEC-10079): as a end user I would like to be able to drag to reposition floating player

### PR DESCRIPTION
**Pull request by RoyBregman**
### Description of the Changes

added dragging functionality - please notice I added also throttling mechanism using setTimeout

Solves FEC-10079

### CheckLists

- [X] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/visibility.js [hunk 1]
```diff
@@ -118,13 +125,20 @@ class Visibility extends BasePlugin {
...
+    }
   }
 
   _startFloating() {
     Utils.Dom.addClassName(this._floatingContainer, FLOATING_ACTIVE_CLASS);
     Utils.Dom.setStyle(this._floatingContainer, 'height', this.config.floating.height + 'px');
     Utils.Dom.setStyle(this._floatingContainer, 'width', this.config.floating.width + 'px');
     Utils.Dom.setStyle(this._floatingContainer, 'margin', `${this.config.floating.marginY}px ${this.config.floating.marginX}px`);
+    if (this.config.floating.draggable) {
+      this.eventManager.listen(this._floatingContainer, 'mousedown', this._dragMouseDown.bind(this));
```
Comment (OrenMe) [inline src/visibility.js:140]: will this also work for touch?

File: src/visibility.js [hunk 2]
```diff
@@ -177,6 +191,45 @@ class Visibility extends BasePlugin {
...
+      const boundClientRect = floatingContainer.getBoundingClientRect();
+      floatingContainer.style.top = boundClientRect.top - parseInt(floatingContainer.style.marginTop) - deltaMousePosY + 'px';
+      floatingContainer.style.left = boundClientRect.left - parseInt(floatingContainer.style.marginLeft) - deltaMousePosX + 'px';
+    }
+
+    // handle throttling to avoid performance issues on dragging
+    this._throttleWait = true;
+    setTimeout(() => {
+      this._throttleWait = false;
+    }, 30);
```
Comment (OrenMe) [inline src/visibility.js:225]: please move magic numbers to const in top of file.
Do we need to set this number as configurable?

File: src/visibility.js [hunk 3]
```diff
@@ -177,6 +191,45 @@ class Visibility extends BasePlugin {
     this._observer = null;
     this.eventManager.destroy();
   }
+
+  _dragMouseDown(e: MouseEvent) {
+    e.preventDefault();
+    // get the mouse cursor position at startup:
+    this._currMousePos.x = e.clientX;
+    this._currMousePos.y = e.clientY;
+    this.eventManager.listen(document, 'mouseup', this._dragStop.bind(this));
```
Comment (OrenMe) [inline src/visibility.js:200]: maybe listen to this once and avoid needing to unregister in event handler

File: src/visibility.js (hunk 1 above)
Comment (RoyBregman) [reply src/visibility.js:140]: Now it is currently not working in mobile. Started yesterday to add touch events but it creates some bugs. We need to decide wether to push it in the sprint or not compared to other tasks

File: src/visibility.js (hunk 2 above)
Comment (RoyBregman) [reply src/visibility.js:225]: This is the miliseconds of the delay. Meaning the the mouse move will hit no more than ~30 times a second.
I will take out as const - don't think it is needed as a config

File: src/visibility.js (hunk 3 above)
Comment (RoyBregman) [reply src/visibility.js:200]: Good idea

File: src/visibility.js [hunk 4]
```diff
@@ -177,6 +191,45 @@ class Visibility extends BasePlugin {
...
+    // calculate the new cursor position:
+    const deltaMousePosX = this._currMousePos.x - e.clientX;
+    const deltaMousePosY = this._currMousePos.y - e.clientY;
+    this._currMousePos.x = e.clientX;
+    this._currMousePos.y = e.clientY;
+    const floatingContainer = this._floatingContainer; // flow
+    // set the element's new position
+    if (floatingContainer) {
+      const boundClientRect = floatingContainer.getBoundingClientRect();
+      floatingContainer.style.top = boundClientRect.top - parseInt(floatingContainer.style.marginTop) - deltaMousePosY + 'px';
```
Comment (Yuvalke) [inline src/visibility.js:217]: I think it's not supported in different browsers - calculation of numbers and concatenate string on the same expression.

File: src/visibility.js (hunk 1 above)
Comment (Yuvalke) [inline src/visibility.js:140]: don't you need to remove [this](https://github.com/kaltura/playkit-js-visibility/pull/9/files#diff-9c3f2b305c531447d257c17efb63ef3cR200) and [this](https://github.com/kaltura/playkit-js-visibility/pull/9/files#diff-9c3f2b305c531447d257c17efb63ef3cR201) listeners on `mouseup` of `_floatingContainer`?
you keep listening to changes in document events when you don't need to.

File: src/visibility.js (hunk 4 above)
Comment (Yuvalke) [reply src/visibility.js:237]: @RoyBregman this expression isn't working in IE11

File: src/visibility.js [hunk 5]
```diff
@@ -177,6 +199,57 @@ class Visibility extends BasePlugin {
...
+  }
+  _clientY(e: any): number {
+    return e.clientY || (e.changedTouches && e.changedTouches[0] && e.changedTouches[0].clientY);
+  }
+
+  _moveDrag(e: any) {
+    if (this._throttleWait) return;
+
+    e = e || window.event;
+    // e.preventDefault();
```
Comment (Yuvalke) [inline src/visibility.js:227]: remove comment line

File: src/visibility.js (hunk 4 above)
Comment (RoyBregman) [reply src/visibility.js:237]: It is supported

File: src/visibility.js (hunk 5 above)
Comment (RoyBregman) [reply src/visibility.js]: removed

File: src/visibility.js [hunk 6]
```diff
@@ -192,24 +200,35 @@ class Visibility extends BasePlugin {
...
+  }
+  _clientX(e: any): number {
+    return e.clientX || (e.changedTouches && e.changedTouches[0] && e.changedTouches[0].clientX);
+  }
+  _clientY(e: any): number {
+    return e.clientY || (e.changedTouches && e.changedTouches[0] && e.changedTouches[0].clientY);
   }
 
-  _floatingDrag(e: MouseEvent) {
+  _moveDrag(e: any) {
```
Comment (OrenMe) [inline src/visibility.js:223]: can we make this type stricter? MouseEvent | Event?

Comment (RoyBregman) [reply src/visibility.js]: In some i changed (pushed it now) and some i couldn't

**Pull request merged**